File: SARAssist/RowTableViewController.m
//
//  RowTableViewController.m
//  SARAssist
//
//  Created by V-FEXrt on 7/18/15.
//  Copyright (c) 2015 V-FEXrt. All rights reserved.
//

#import "RowTableViewController.h"
#import "RowTableViewCell.h"
#import <Parse/Parse.h>

#import "MapViewController.h"
@interface RowTableViewController ()
@property (nonatomic, strong)NSArray *blockRows;
@property (nonatomic, strong)PFGeoPoint *currentLocation;
@end

@implementation RowTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [PFGeoPoint geoPointForCurrentLocationInBackground:^(PFGeoPoint *geoPoint, NSError *error) {
        if (geoPoint) {
            //Save the current location
            self.currentLocation = geoPoint;
            
            PFQuery *query = [PFQuery queryWithClassName:@"Block"];
            
            [query whereKey:@"SearchAreaID" equalTo:self.selectedArea[@"SearchAreaID"]];
            
            [query findObjectsInBackgroundWithBlock:^(NSArray *objects, NSError *error) {
                if (!error) {
                    
                    NSMutableArray *arr = [[NSMutableArray alloc]init];
                    
                    int max = -1;
                    for (PFObject *obj in objects) {
                        if ([obj[@"Row"] intValue] > max) {
                            max = [obj[@"Row"]intValue];
                        }
                    }
                    
                    //row starts at 0, therefore max is one less than expected
                    for (int i = 0; i <= max; i++) {
                        NSMutableArray *obj = [[NSMutableArray alloc]init];
                        [arr addObject:obj];
                    }
                    
                    for (PFObject *obj in objects) {
                        int index = [obj[@"Row"]intValue];
                        [arr[index] addObject:obj];
                    }
                    
                    self.blockRows = [arr copy];
                    
                    [self.tableView reloadData];
                    
                } else {
                    // Log details of the failure
                    NSLog(@"Error: %@ %@", error, [error userInfo]);
                }
            }];
            
        }
        else
        {
        }
        
    }];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.blockRows) {
        return [self.blockRows count];
    }
    return 0;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    RowTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    //Get the first object in the current row
    PFObject *obj = [self.blockRows[indexPath.row]firstObject];
    
    if (obj[@"AssignedTo"] && ![obj[@"AssignedTo"] isEqualToString:@""]) {
        cell.title.text = obj[@"AssignedTo"];
        cell.subtitle.text = obj[@"updatedAt"];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        
    }else{
        cell.title.text = @"Not Assigned";
        double distance = [self.currentLocation distanceInMilesTo:obj[@"Location"]];
        cell.subtitle.text = [NSString stringWithFormat:@"%.1f miles away", distance];
    }

    
    // Configure the cell...
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Select Row?" message:@"Are you sure you want to select this row?" preferredStyle:UIAlertControllerStyleActionSheet];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Select Row" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
        
        
        [self performSegueWithIdentifier:@"segue" sender:self];
        
    }]];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {

    }]];
    
    [self presentViewController:alertController animated:true completion:nil];
}


#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
#warning Need to mark blocks as selected
    NSIndexPath *path = [self.tableView indexPathForSelectedRow];
    
    MapViewController *controller = [segue destinationViewController];
    
    controller.selectedBlocks = self.blockRows[path.row];
}


@end

File: SARAssist_Swift/SARAssist/RowTableViewController.swift
//
//  RowTableViewController.swift
//  SARAssist
//
//  Created by V-FEXrt on 9/18/15.
//  Copyright (c) 2015 V-FEXrt. All rights reserved.
//

import UIKit

import Parse

class RowTableViewController: UITableViewController {

    internal var selectedArea:AreaModel = AreaModel()
    internal var currentLocation:PFGeoPoint = PFGeoPoint()
    
    private var blockRows:Array2DTyped<BlockModel> = Array2DTyped(cols: 0, rows: 0, defaultValue: BlockModel())
    
    private let formatter = NSDateFormatter()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.formatter.dateStyle = .ShortStyle
        self.formatter.timeStyle = .FullStyle
        
        let query:PFQuery = BlockModel.query()!
        
        query.whereKey("SearchAreaID", equalTo: self.selectedArea.SearchAreaID)
        query.orderByAscending("Column")
        query.limit = 1000
        
        query.findObjectsInBackgroundWithBlock { (objects:[PFObject]?, error:NSError?) -> Void in
            if let obj:[AnyObject] = objects{
                let blocks:[BlockModel] = (obj as? [BlockModel])!
                
                var maxRow:Int = -1
                var maxCol:Int = -1
                
                
                for block:BlockModel in blocks{
                    if(block.Row.integerValue > maxRow){
                        maxRow = block.Row.integerValue
                    }
                    if(block.Column.integerValue > maxCol){
                        maxCol = block.Column.integerValue
                    }
                }
                
                self.blockRows = Array2DTyped(cols: maxCol+1, rows: maxRow+1, defaultValue: BlockModel())
                
                for block:BlockModel in blocks{
                    self.blockRows[block.Column.integerValue, block.Row.integerValue] = block
                }
                
                self.tableView.reloadData()
            }
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete method implementation.
        // Return the number of rows in the section.
        return self.blockRows.rowCount()
    }

    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        let cell:RowTableViewCell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath) as! RowTableViewCell

        
        //Get the first object in the current row
        let block:BlockModel = self.blockRows[indexPath.row, 0]
        
        
        if ((block.AssignedTo) != nil) {
            let user:PFUser = block.AssignedTo;
            
            
            user.fetchIfNeededInBackgroundWithBlock({ (obj:PFObject?, error:NSError?) -> Void in
                cell.title.text = user.username
                cell.subtitle.text = self.formatter.stringFromDate((obj?.updatedAt)!)
                //cell.selectionStyle = .None
                
            })

        }else{
            cell.title.text = "Not Assigned";
            let distance:Double = self.currentLocation.distanceInMilesTo(block.Location)
            
            cell.subtitle.text = String(format: "%.1f miles away", distance )
        }


        return cell
    }
    

    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        
        let cell:RowTableViewCell = self.tableView.cellForRowAtIndexPath(indexPath) as! RowTableViewCell
        if(cell.title.text == "Not Assigned" || cell.title.text == PFUser.currentUser()?.username){
            
            let alertController:UIAlertController = UIAlertController(title: "Select Row?", message: "Are you sure you want to select this row?", preferredStyle: .ActionSheet)
            
            if(UI_USER_INTERFACE_IDIOM() == .Pad){
                alertController.popoverPresentationController?.sourceView = self.view
                
                alertController.popoverPresentationController!.sourceRect = CGRectMake((cell.bounds.origin.x + cell.title.bounds.size.width), cell.bounds.origin.y, 20, 20);
                
            }
            
            alertController.addAction(UIAlertAction(title: "Select Row", style: UIAlertActionStyle.Default, handler: { (UIAlertAction) -> Void in
                self.performSegueWithIdentifier("segue", sender: self)
            }))
            
            alertController.addAction(UIAlertAction(title: "Cancel", style: UIAlertActionStyle.Cancel, handler: nil))
            
            self.presentViewController(alertController, animated: true, completion: nil)
        }

    }

    
    // MARK: - Navigation


    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {

        let path:NSIndexPath = self.tableView.indexPathForSelectedRow!
        
        for block:BlockModel in self.blockRows[path.row]{
            
            block.AssignedTo = PFUser.currentUser()
            block.saveInBackgroundWithBlock({ (success:Bool, error:NSError?) -> Void in
                if(success){
                    let cell:RowTableViewCell = self.tableView.cellForRowAtIndexPath(path) as! RowTableViewCell
                    cell.title.text = PFUser.currentUser()?.username
                    cell.subtitle.text = self.formatter.stringFromDate(block.updatedAt!)
                }
            })
        }
        
        let vc:MapViewController = segue.destinationViewController as! MapViewController
        
        vc.selectedBlocks = self.blockRows[path.row];
        vc.currentLocation = self.currentLocation
    }


}
